src/plugins/data/public/ui/index_pattern_select/index_pattern_select.tsx

```diff
@@ -107,39 +102,59 @@ export default class IndexPatternSelect extends Component<IndexPatternSelectInte
   };
 
   debouncedFetch = _.debounce(async (searchValue: string) => {
-    const { fieldTypes, onNoIndexPatterns, indexPatternService } = this.props;
-    const indexPatterns = await indexPatternService.find(
-      `${searchValue}*`,
-      this.props.maxIndexPatterns
-    );
+    const isCurrentSearch = () => {
+      return this.isMounted && searchValue === this.state.searchValue;
+    };
 
-    // We need this check to handle the case where search results come back in a different
-    // order than they were sent out. Only load results for the most recent search.
-    if (searchValue !== this.state.searchValue || !this.isMounted) {
+    const idsAndTitles = await this.props.indexPatternService.getIdsWithTitle();
+    if (!isCurrentSearch()) {
       return;
     }
 
-    const options = indexPatterns
-      .filter((indexPattern) => {
-        return fieldTypes
-          ? indexPattern.fields.some((field) => {
-              return fieldTypes.includes(field.type);
-            })
-          : true;
-      })
-      .map((indexPattern) => {
-        return {
-          label: indexPattern.title,
-          value: indexPattern.id,
-        };
+    const options = [];
+    for (let i = 0; i < idsAndTitles.length; i++) {
+      if (!idsAndTitles[i].title.toLowerCase().includes(searchValue.toLowerCase())) {
+        // index pattern excluded due to title not matching search
+        continue;
+      }
+
+      if (this.props.fieldTypes) {
+        try {
+          const indexPattern = await this.props.indexPatternService.get(idsAndTitles[i].id);
+          if (!isCurrentSearch()) {
+            return;
+          }
+          const hasRequiredFieldTypes = indexPattern.fields.some((field) => {
+            return this.props.fieldTypes!.includes(field.type);
+          });
+          if (!hasRequiredFieldTypes) {
+            continue;
+          }
+        } catch (err) {
+          // could not load index pattern, exclude it from list.
+          continue;
+        }
+      }
+
+      options.push({
+        label: idsAndTitles[i].title,
+        value: idsAndTitles[i].id,
       });
+
+      // Loading each index pattern object requires a network call so just find small number of matching index patterns
+      // Users can use 'searchValue' to further refine the list and locate their index pattern.
+      if (options.length > 15) {
+        break;
+      }
+    }
+
     this.setState({
       isLoading: false,
       options,
     });
 
-    if (onNoIndexPatterns && searchValue === '' && options.length === 0) {
-      onNoIndexPatterns();
+    if (this.props.onNoIndexPatterns && searchValue === '' && options.length === 0) {
```

Comment:
Does `searchValue` have to be an empty string here? Couldn't there just be no returned index patterns for any search value?

Reply:
This case is the specific case where there are no index patterns or no index patterns matching the provided field types. In this case, the UI displays a message like "hey, you don't have any index patterns, set them up". Its important to not conflate this case of no index patterns with the case where there are no index patterns because of the search string. In the later, where there are no index patterns because of the search string, you do not want to show the "no index patterns - make some" message because there are index patterns, they were just filtered out by the search string.